### PR TITLE
Add support for custom shells/editors

### LINKED ISSUE #5083: Allow adding custom editor commands as "External Editor" option
**Is your feature request related to a problem? Please describe.**
Editor-specific implementation like https://github.com/desktop/desktop/issues/4785 aside, users may have specific expecations of how their favorite editor is invoked. Their needs can never be fulfilled by adding explicit support for editors.

In the case of emacs for instance,
- A user with a lightweight Emacs configuration will prefer to use the ``emacs`` executable, or ``runemacs.exe`` on Windows.
- An advanced Emacs user with a heavy configuration, causing the startup process to take a significant amount of time, will likely prefer some invocation of ``emacsclient``, but depending on operating system and preferences the details may vary, e.g.
    - ``emacsclient -a emacs <dirname>`` to handle cases when no emacs session is running a server yet.
    - ``emacsclient -a "emacs -Q" <dirname>`` if the user prefers quick startup over complete startup, if no emacs server is running yet.
    - ``<some wrapper script> <dirname>`` if the user has a more complicated setup.

**Describe the solution you'd like**
Adding custom command strings as *External Editor* option. In the command string, a pattern like ``%d`` should be replaced by the full directory path of the repository.

**Describe alternatives you've considered**
None really. Switching editors is not viable, and only a customizable command can fit my usage.

**Teachability, Documentation, Adoption, Migration Strategy**
A mockup says more than a thoudsand words.

### PR DESCRIPTION
**Pull request by sergiou87**
Closes #5083
Closes https://github.com/github/desktop/issues/830

## Description

This PR adds support for custom editors and shells. This feature allows users to select a custom editor or shell to open repositories in. The user can choose from a list of predefined editors and shells or specify a custom editor or shell. The user can also specify command line arguments to pass to the custom editor or shell when opening a repository.

In order to parse command line arguments following Windows rules, this PR includes a new internal/private node addon which uses the Windows `CommandLineToArgvW` API, while the addon `string-to-argv` is used for the same purpose on UNIX-based systems.

In an upcoming PR, we will revisit how and when changes in preferences are validated: both when switching to a different section in preferences and when clicking on the `Save` button.

Until then, these changes are feature flagged to `development` builds only.

### Screenshots

image

image

image

image

image

## Release notes

Notes: [New] Add support for custom editors and shells

### CODE REVIEW

**Comment by StuartJAtkinson** [comment]
Looking forward to this I'm hoping to tie in Cursor as my editor.

**Review by StuartJAtkinson — COMMENTED** [review]
First review of something, this looks fine it seems to correct the backend function. Although it doesn't seem to change the UI to allow the selection of the custom shell/editor?

**Comment by tidy-dev** [comment]
Just trying out the UI (not looking at code yet). The functionality is great. (I just used the notes app and it surprising opens a note and adds the file as an attachment... if you open a repo, it does a file import 🤯 ) 

Feedback TLDR:
1. Find a way to not have the error show on form appearance.
2. Put whitespace between shell and editor sections to break up the content some. (or otherwise simply the form entry) It even makes me think of something that fieldsets are nice for since they are distinctly different contexts that you may not mess with one while messing with the other.

My first thoughts are that when using other on both.. the screen is very overwhelming. Below is my brainstorming.. so maybe worth looking at maybe something with equal but just different pros/cons. :P  Also, the "other" select has it's accessibility qualms because we are hiding and showing inputs with out using a "control" that can be described with aria-labeling to inform a user that a form will be appearing, but it is an established pattern in our app so I am not firmly opposed to it. Just looking to see if we can do better in a user friendly way.

1. Form  Validation
This form validation is a different pattern then our other forms. In other forms in the app we have the submit button disabled until valid input instead of immediately showing a validation error before you get to attempt to put something in. Can we do that here to prevent being pestered before a chance to do something about it?

"submit button disabled" is an accessibility anti-pattern that one of our accessibility issues is tracking. But... until we switch all our dialogs, maybe we should continue that expected submission pattern? But for reference... A friendlier accessibility pattern would be to keep the save button enabled, run a validation check that then shows the validation error beneath the input, and (very importantly) moves the users focus into the errored input. Which.. I believe some of the save dialog already lends that pattern because we have a "onSave" method that would be a likely place to put this logic unless solving the anti-pattern for dialogs included adding some automatic triggering/handling of validation flow on dialog submit. 

2. Simplify form thoughts (which could in turn help with 1)
What if configuring your custom shell or editor was done in another dialog?  This feels like the git ignore discussion again.. but... brainstorming ... 

What if beneath the input we had a "Don't see your editor? [Configure a custom editor]". When you click on that it opens a dialog just for finding and adding our path/arguments. Then, on submit you return to the integration settings. And you have in plain text (multiple lines). Path: ..... Arguments: ... A button beneath that says [ Configure a custom editor].  The dialog approach would help this screen be less busy/overwhelming by simplifying what you must interact with at a given time and would simply validation down to just this input.  As for the editor select input.. could keep the selection input with "Other" bit, or have a [Use detected editor] (maybe even have the select there but disabled?) that returns it to the previous state (but I would remember the users last editor so if the clicked configure a custom editor we could pre-populate to their last selection). 

I think a clear advantage to your current design is if a user wants to toggle between "VS code" and "custom editor". Switching between those in the select is quick. My assumption is tho that people don't do that a lot (especially for shell integration) - i.e. if they have an unsupported editor, they will go into this dialog flow once and hit save.

Another pattern we have throughout the app is radio buttons.
- Use detected editor
      -  Enabled/Disabled select input
- Configure editor path.
    [either form input, or same dialog approach]

Or... really spitballing.
What if these are both buttons. [Use detected editor] [Configure editor path] .. both open a dialog.. K maybe that is getting into a lot of clicks for the detected editor path. 

Or.. tabs [Use detected editor] [Configure editor path] .. at least from accessibility that would clearly transition that a new form may appear.. that the select input does not.

**Comment by sergiou87** [comment]
> Find a way to not have the error show on form appearance

Sounds good! I think we could even just show the "open file" dialog when the user clicks on `Other…` for the first time, when a custom editor/shell wasn't previously configured.

> A friendlier accessibility pattern would be to keep the save button enabled, run a validation check that then shows the validation error beneath the input, and (very importantly) moves the users focus into the errored input

👍  FWIW I followed the pattern of "Other email" in the `Git` config form to display errors as the user types. In terms of what happens when `Save` is clicked, if you enter an invalid "default branch name" (in the `Git` section) and click on `Save`, it won't complain and just won't save that change. But I can do the form validation on save and focus on the right input before closing the dialog!

> Simplify form thoughts

I agree it's overwhelming, but I also think it's an "advanced" setting so it shouldn't be _that bad_ if we couldn't come up with a better approach. I also agree with separating both more, but maybe using `h2` instead of `h3` is enough? (And removing the `Applications` title) :
image
image

Moving this form to a new dialog or using radio buttons feels like a more unnecessarily "convoluted" UX. I'm not extremely opposed to any of that, but _to me_ it feels like the current implementation is simpler (despite how ugly/overwhelming/complex it already is 😕 ) and uses a more standard pattern 🤔

**Comment by tidy-dev** [comment]
> using h2 instead of h3

This does help a lot! 

 I was just playing with what if we but margin-top on all h2 across the setting dialog (except for first one or.. having each "setting section" have margin-bottom .. whichever is most straight forward to implement.). What do you think? I like it. 

Showing the h2's having margin-top on integrations page

Showing the h2's having margin-top on other setting page

**Comment by tidy-dev** [comment]
I think the select input now needs a label since path and arguments have one that isn't the heading. I know we link the heading to technically fit accessibility criteria, but for visual users this feels like the select input is missing a label.

**Comment by sergiou87** [comment]
I like that! Thanks 💖 

> I think the select input now needs a label since path and arguments have one that isn't the heading. I know we link the heading to technically fit accessibility criteria, but for visual users this feels like the select input is missing a label.

Don't you think the heading is visually close enough to the input so that it feels like its label? At least, that's how I see it. It's rendered the same way as a label, but with a bigger bold font 🤔

**Comment by tidy-dev** [comment]
> just show the "open file" dialog when the user clicks on Other… for the first time

This feels confusing to me, because choosing a select option usually isn't something opens a dialog (like clicking a button).

This did get me thinking, maybe the word shouldn't be "Other..", but "Configure editor..." so it is more clear that selecting it will lead to doing something (a screen change) and what that something is..  I think that would help with accessibility qualms around hiding/showing form elements. 

Other thoughts, if we think people will be generally use the "open file" dialog, shouldn't that be the first element.. will people even use the text input? I know it doesn't follow our other choose file patterns we have in the app tho so maybe not a good idea... but in the other places, I think we generally pre-populate the path for the users because it is referring to repository paths so the choose button is more secondary.

**Comment by tidy-dev** [comment]
> Don't you think the heading is visually close enough to the input so that it feels like its label?

:D This is a kind of pet peeve I have with our settings dialog. We don't have a consistent convention.  In my opinion, the only way for us to be consistent is to use headers as headers and each input have a label. I think the integrations page before did have a header (Applications) and then we used h3 headers as labels that could have really been just labels. 

Here no heading above the first two inputs... no label above the bottom input. (I know the heading is.. but that is different than the above inputs so it feels off to me).
No heading above inputs

Heading + label. 
Here we have heading plus label.

The other settings generally are checkboxes or radios which are headings plus individual labels. Could be argued that they are a little bit of a blur because the headings are radio group labels too.. but each input still has it's own clear visual label. 

In other dialogs in our app, we pretty consistently provide labels for each input.
Showing create repository

**Comment by sergiou87** [comment]
> This did get me thinking, maybe the word shouldn't be "Other..", but "Configure editor..." so it is more clear that selecting it will lead to doing something (a screen change) and what that something is.. I think that would help with accessibility qualms around hiding/showing form elements.

Isn't the `…` in `Other…` already the hint to indicate "there is something more behind this action/option/menu"?

> but each input still has it's own clear visual label

Just to be clear: my argument is that in this specific case, that heading is the label (it's positioned in the same place as a label would, but slightly different style - bigger font). I think adding an additional label in this case makes the whole form more "crammed" and doesn't add much value/info 😳

**Comment by tidy-dev** [comment]
> Isn't the … in Other… already the hint to indicate "there is something more behind this action/option/menu"?

Yes. I agree it communicates something will happen. But, not the "what". If I do not know that I can configure an editor and since a select input would indicate that there isn't a way to do it since it indicates that I have a predefined list to choose from (more reasons this pattern is not accessibility/user-friendly). If I open the select options and see "Other..." vs "Configure editor...", I would have much more confidence in what I am about to do.  This kind of falls in that realm of why we don't like "Learn more..." links, it doesn't communicate what is going to happen and if I wan't to bother clicking there. It is more user friendly to communicate upfront. This is why I like the idea of a button beneath the select for "Configure custom editor". 

For more visual. 
**External editor**
Detected Editors
[ select list of predefined options] 

[ Configure custom editor] (button) 

If I came to something like above, I would immediately know I can choose a predefined option or configure a custom editor. 

>  doesn't add much value/info

That I can agree with and think that the label wording would be difficult to choose. (like do we just repeat "external editor", or try to be more direct like "Detected editors"

> makes the whole form more "crammed"

My opinion is it looks funny - like a pixel being off - the heading looks like a heading and it looks like the input  doesn't have a label to me. It takes me a second longer to digest that oh that is probably an editor list because of the heading.

**Comment by tsvetilian-ty** [comment]
@sergiou87 great implementation 👏  I have a question regarding `string-to-argv` Is there a reason you picked `string-to-argv` instead of `string-argv`?

File: app/src/ui/preferences/custom-integration-form.tsx [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    )
+  }
+
+  private renderPathErrors() {
+    if (this.state.isValidPath || !this.state.showNonValidPathWarning) {
+      return null
+    }
+
+    const errorDescription =
+      'This directory does not appear to be a valid executable.'
```
Comment (tsvetilian-ty) [inline app/src/ui/preferences/custom-integration-form.tsx]: @sergiou87 doesn't it make more sense for the error to be: `This path does not appear to be a valid executable.`?

Comment (sergiou87) [reply app/src/ui/preferences/custom-integration-form.tsx]: 🤦‍♂️  Yes, thank you! 🙇‍♂️

**Comment by sergiou87** [comment]
@tsvetilian-ty it seems I made a mistake with `string-to-argv`: I looked at several options, but it still took a few weeks until I actually added the dependency to the project and confused `string-argv` with `string-to-argv` 🤦‍♂️ 

I will fix it, thank you again, nice catch! 🙏

File: app/src/ui/preferences/custom-integration-form.tsx (hunk 1 above)
Comment (sergiou87) [reply app/src/ui/preferences/custom-integration-form.tsx]: Fixed in [8915043](https://github.com/desktop/desktop/pull/18954/commits/8915043c76e9654a962e96b5091ee5c22e53a90a), thanks again 💖

**Review by tidy-dev — APPROVED** [review]
✨  This is amazing work!

Since you already added the suggested next editor bit ✨, my only thought while using it was that the menu items that say "Open in External Editor" or "Open in Shell" should say "Open in Custom Editor" and "Open in Custom Shell" for a higher signal that I am using what I set up or just so I know what I am going to open. Even better let users provide a reference name like we do for provided editors (I know.. that form is already too crowded.. :/ but I think it would make the experience a bit nicer). If you think this is worth exploring, it would just make a nice iteration (not a must have for this PR)

**Comment by sergiou87** [comment]
I will take a look in a different PR, but it sounds good, thank you!!

**Comment by Cedric-Ab** [comment]
How would one know what variables are available?

**Comment by steveward** [comment]
> How would one know what variables are available?

That would depend on the editor, but you would want to look at their documentation. For instance, Unity has a list of arguments available:

https://docs.unity3d.com/Manual/EditorCommandLineArguments.html

**Comment by Cedric-Ab** [comment]
> > How would one know what variables are available?
> 
> That would depend on the editor, but you would want to look at their documentation. For instance, Unity has a list of arguments available:
> 
> https://docs.unity3d.com/Manual/EditorCommandLineArguments.html

My question is more if GitHub Desktop got the ability to Hove more variables than the path?

**Review by skclabs — COMMENTED** [review]
CRJ INTEGRATION

**Comment by Coinsintegrity** [comment]
> Closes #5083 Closes https://github.com/github/desktop/issues/830
> 
> ## Description
> This PR adds support for custom editors and shells. This feature allows users to select a custom editor or shell to open repositories in. The user can choose from a list of predefined editors and shells or specify a custom editor or shell. The user can also specify command line arguments to pass to the custom editor or shell when opening a repository.
> 
> In order to parse command line arguments following Windows rules, this PR includes a new internal/private node addon which uses the Windows `CommandLineToArgvW` API, while the addon `string-to-argv` is used for the same purpose on UNIX-based systems.
> 
> In an upcoming PR, we will revisit how and when changes in preferences are validated: both when switching to a different section in preferences and when clicking on the `Save` button.
> 
> Until then, these changes are feature flagged to `development` builds only.
> 
> ### Screenshots
> image
> 
> image
> 
> image
> 
> image
> 
> image
> 
> ## Release notes
> Notes: [New] Add support for custom editors and shells

https://github.com/desktop/desktop/pull/18954#issue-2402771183/mairo1504@gmail.com

**Pull request merged**